### PR TITLE
Move mimetypes.list.php to config/mimetypemapping.dist.json

### LINKED ISSUE #15384: Move mimetypes.list.php to a text file in /config
It can take between 1 to 6+ months for a user to have access to a new media type, simply because `core` has to be hacked for that.

I propose we move the list to a simple text or php file in the config folder so that users can upgrade the list as they see fit.
The cron job can keep the cache updated and simple media type migrations could be performed on the fly.

Is there some big invisible hurdle to doing it this way?

### PR DESCRIPTION
**Pull request by rullzer**
This allows users to add new mimetypemappings (extention -> mimetype)
themself. And not have to wait until a new release for updated
mimetypes.

Fixes: #15384 

CC: @oparoz @Xenopathic @MorrisJobke @PVince81 @DeepDiver1975

### CODE REVIEW

**Comment by rullzer** [comment]
mmmm, we actually should make sure that this file does not get overwritten during an upgrade... else the user loses his additions...
Same actually holds for config/mimetypealiases.json

**Comment by oparoz** [comment]
Isn't the config folder left untouched?

**Comment by Xenopathic** [comment]
Perhaps we should support loading multiple files, e.g. `mimetypemapping.custom.json`, which is packaged as empty and never gets overridden by updates. Any entries in that file override the ones in `mimetypemapping.json`, which has a note at the top explicitly telling admins not to put stuff there.

Or do it the other way around - make the custom one `mimetypemapping.json`, with a `mimetypemapping.dist.json` for the distributed packaged one.

**Comment by Xenopathic** [comment]
But otherwise this PR looks good :+1:

**Comment by oparoz** [comment]
Good call @Xenopathic. Makes it clear what is what.

**Comment by Xenopathic** [comment]
Interesting side note, oC actually loads any config files with the name `*.config.php` (https://github.com/owncloud/core/blob/master/lib/private/config.php#L183). Not particularly useful for configs, but very handy in this case where we need a default packaged file that gets updated.

**Comment by rullzer** [comment]
Agreed lets use mimetypemapping.dist.json... let me fix that for this PR. And create a new one for the mimetypealiases.json

**Comment by rullzer** [comment]
Done :)

File: config/mimetypemapping.dist.json [hunk 1]
```diff
@@ -0,0 +1,170 @@
+{
+	"_comment" : "Array mapping file extensions to mimetypes (in alphabetical order]",
+	"_comment2": "The first index in the mime type array is the assumed correct mimetype",
+	"_comment3": "and the second (if present] is a secure alternative",
+
+	"_comment4": "To add your own mappings create a file mimetypemapping.json",
```
Comment (Xenopathic) [inline config/mimetypemapping.dist.json]: This needs to be phrased like a warning:

> Any changes you make here will be overwritten on an update of ownCloud
> Put any custom mappings in a new file mimetypemapping.json in this directory

Comment (rullzer) [inline config/mimetypemapping.dist.json]: Done :)

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
@owncloud-bot retest this please

**Comment by Xenopathic** [comment]
Nice! :+1:

We need some documentation for this though. And do we want to ship a blank `mimetypemapping.json` with only comments saying how to use the file? Or do we want to require the user to manually create the file, and be expected to know the syntax etc?

**Comment by oparoz** [comment]
@Xenopathic I'd say it's an advanced feature and it shouldn't be too much to ask of an admin to create the file.

**Comment by rullzer** [comment]
Added todo issue for the docs.

One more review guys :)

File: lib/private/helper.php [hunk 2]
```diff
@@ -424,8 +424,18 @@ static function rmdirr($dir, $deleteSelf = true) {
 	 */
 	static public function getMimetypeDetector() {
 		if (!self::$mimetypeDetector) {
+			$dist = file_get_contents(OC::$SERVERROOT . '/config/mimetypemapping.dist.json');
```
Comment (MorrisJobke) [inline lib/private/helper.php]: @rullzer Can you change this to `\OC::$configDir`?

Comment (rullzer) [inline lib/private/helper.php]: Ah didn't know about that one. Fixed!

**Comment by rullzer** [comment]
@owncloud-bot retest this please

File: lib/private/helper.php [hunk 3]
```diff
@@ -424,8 +424,18 @@ static function rmdirr($dir, $deleteSelf = true) {
 	 */
 	static public function getMimetypeDetector() {
 		if (!self::$mimetypeDetector) {
+			$dist = file_get_contents(OC::$configDir . '/mimetypemapping.dist.json');
+			$mimetypemapping = get_object_vars(json_decode($dist));
+
+			//Check if need to load custom mappings
+			if (file_exists(OC::$configDir . '/mimetypemapping.json')) {
+				$custom = file_get_contents(OC::$configDir . '/mimetypemapping.json');
```
Comment (Xenopathic) [inline lib/private/helper.php]: Actually, TBH I think the code looks better with those slashes there. PHP ignores multiple slashes (as does any UNIX utility), so we have no bug concerns, and it also means less effort if `$configDir` at some point loses the trailing slash.

**Comment by Xenopathic** [comment]
:+1:

**Comment by rullzer** [comment]
@owncloud-bot retest this please

**Comment by Xenopathic** [comment]
@owncloud-bot retest this please

**Comment by MorrisJobke** [comment]
Mime type detection still works :+1:

**Pull request merged**